relayer/src/supervisor/spawn.rs

```diff
@@ -514,22 +508,48 @@ impl<'a> SpawnContext<'a> {
                 )
                 .then(|| debug!("spawned Client worker: {}", client_object.short_name()));
 
-            // create the Packet object and spawn worker
-            let path_object = Object::Packet(Packet {
-                dst_chain_id: counterparty_chain.id(),
-                src_chain_id: chain.id(),
-                src_channel_id: channel.channel_id,
-                src_port_id: channel.port_id,
-            });
+            let counterparty_channel = IdentifiedChannelEnd::new(
+                channel.channel_end.counterparty().port_id.clone(),
+                channel
+                    .channel_end
+                    .counterparty()
+                    .channel_id
+                    .clone()
+                    .unwrap(),
+                counterparty_channel.unwrap(), // safe because chan_state_dst is Open
+            );
 
-            self.workers
-                .spawn(
-                    chain.clone(),
-                    counterparty_chain.clone(),
-                    &path_object,
-                    &self.config.read().expect("poisoned lock"),
-                )
-                .then(|| debug!("spawned Path worker: {}", path_object.short_name()));
+            let outstanding_packets = !unreceived_packets(
+                counterparty_chain.as_ref(),
+                chain.as_ref(),
+                counterparty_channel.clone(),
+            )?
+            .is_empty()
+                || !unreceived_acknowledgements(
+                    counterparty_chain.as_ref(),
+                    chain.as_ref(),
+                    counterparty_channel.clone(),
+                )?
+                .is_empty();
```

Comment:
I find that this code is not particularly easy to read because of all the nested function calls. Can we add some `let` bindings to give names to the sub-expressions?

Reply:
850ca46d